File: internal/agent/setup/ollama_tools_test.go
package setup

import (
	"strings"
	"testing"

	"github.com/ollama/ollama/api"
	"google.golang.org/adk/model"
	"google.golang.org/genai"
)

func TestGenerationOptions(t *testing.T) {
	opts := generationOptions(nil)
	if opts["temperature"] != 0.0 || opts["num_ctx"] != defaultNumCtx {
		t.Errorf("defaults = %+v", opts)
	}
	temp, top := float32(0.5), float32(0.9)
	var seed int32 = 7
	req := &model.LLMRequest{Config: &genai.GenerateContentConfig{Temperature: &temp, TopP: &top, Seed: &seed}}
	opts = generationOptions(req)
	if opts["temperature"] != 0.5 || opts["seed"] != 7 {
		t.Errorf("honored config = %+v", opts)
	}
	if tp, _ := opts["top_p"].(float64); tp < 0.89 || tp > 0.91 { // float32->float64 artifact
		t.Errorf("top_p = %v", opts["top_p"])
	}
}

func TestWantsJSON(t *testing.T) {
	if wantsJSON(nil) {
		t.Error("nil should not want json")
	}
	req := &model.LLMRequest{Config: &genai.GenerateContentConfig{ResponseMIMEType: "application/json"}}
	if !wantsJSON(req) {
		t.Error("application/json should want json")
	}
}

func TestToOllamaTools(t *testing.T) {
	req := &model.LLMRequest{Config: &genai.GenerateContentConfig{Tools: []*genai.Tool{{
		FunctionDeclarations: []*genai.FunctionDeclaration{{
			Name:        "read_file",
			Description: "read a repo file",
			Parameters: &genai.Schema{
				Type:       genai.TypeObject,
				Properties: map[string]*genai.Schema{"path": {Type: genai.TypeString, Description: "the path"}},
				Required:   []string{"path"},
			},
		}},
	}}}}
	tools := toOllamaTools(req)
	if len(tools) != 1 {
		t.Fatalf("tools = %+v", tools)
	}
	fn := tools[0].Function
	if fn.Name != "read_file" || tools[0].Type != "function" {
		t.Errorf("fn = %+v", fn)
	}
	if fn.Parameters.Type != "object" || len(fn.Parameters.Required) != 1 {
		t.Errorf("params = %+v", fn.Parameters)
	}
	prop, ok := fn.Parameters.Properties.Get("path")
	if !ok || len(prop.Type) != 1 || prop.Type[0] != "string" {
		t.Errorf("path prop = %+v (ok=%v) — type must be lowercased", prop, ok)
	}
	if toOllamaTools(&model.LLMRequest{}) != nil {
		t.Error("no config -> nil tools")
	}
}

func TestFinalResponseWithToolCalls(t *testing.T) {
	args := api.NewToolCallFunctionArguments()
	args.Set("city", "Paris")
	tc := api.ToolCall{Function: api.ToolCallFunction{Name: "get_weather", Arguments: args}}

	resp := finalResponse("here you go", []api.ToolCall{tc}, "gemma4")
	if !resp.TurnComplete || resp.FinishReason != genai.FinishReasonStop {
		t.Error("final response should be complete")
	}
	if len(resp.Content.Parts) != 2 {
		t.Fatalf("parts = %+v", resp.Content.Parts)
	}
	if resp.Content.Parts[0].Text != "here you go" {
		t.Errorf("text part = %q", resp.Content.Parts[0].Text)
	}
	fc := resp.Content.Parts[1].FunctionCall
	if fc == nil || fc.Name != "get_weather" || fc.Args["city"] != "Paris" {
		t.Errorf("function call = %+v", fc)
	}

	// no tool calls + no text -> empty parts, still complete
	if r := finalResponse("", nil, "gemma4"); len(r.Content.Parts) != 0 {
		t.Errorf("empty = %+v", r.Content.Parts)
	}
}

func TestToOllamaMessagesToolRoundTrip(t *testing.T) {
	req := &model.LLMRequest{
		Config: &genai.GenerateContentConfig{SystemInstruction: genai.NewContentFromText("sys", genai.RoleUser)},
		Contents: []*genai.Content{
			{Role: genai.RoleUser, Parts: []*genai.Part{{Text: "read a.go"}}},
			{Role: genai.RoleModel, Parts: []*genai.Part{{FunctionCall: &genai.FunctionCall{Name: "read_file", Args: map[string]any{"path": "a.go"}}}}},
			{Role: genai.RoleUser, Parts: []*genai.Part{{FunctionResponse: &genai.FunctionResponse{Name: "read_file", Response: map[string]any{"content": "package a"}}}}},
		},
	}
	msgs := toOllamaMessages(req)
	if len(msgs) != 4 {
		t.Fatalf("msgs = %+v", msgs)
	}
	if msgs[0].Role != "system" || msgs[1].Role != "user" {
		t.Errorf("roles = %q %q", msgs[0].Role, msgs[1].Role)
	}
	if msgs[2].Role != "assistant" || len(msgs[2].ToolCalls) != 1 || msgs[2].ToolCalls[0].Function.Name != "read_file" {
		t.Errorf("assistant tool call = %+v", msgs[2])
	}
	if msgs[3].Role != "tool" || msgs[3].ToolName != "read_file" || !strings.Contains(msgs[3].Content, "package a") {
		t.Errorf("tool result msg = %+v", msgs[3])
	}
}

func TestToGenaiFunctionCall(t *testing.T) {
	args := api.NewToolCallFunctionArguments()
	args.Set("x", "y")
	fc := toGenaiFunctionCall(api.ToolCall{Function: api.ToolCallFunction{Name: "f", Arguments: args}})
	if fc.Name != "f" || fc.Args["x"] != "y" {
		t.Errorf("fc = %+v", fc)
	}
}

func TestJSONString(t *testing.T) {
	if jsonString(nil) != "" {
		t.Error("nil -> empty")
	}
	if got := jsonString(map[string]any{"a": "b"}); got != `{"a":"b"}` {
		t.Errorf("got %q", got)
	}
}

File: kotlin/src/test/kotlin/com/automation/agent/agent/setup/OllamaToolsTest.kt
package com.automation.agent.agent.setup

import com.google.adk.kt.models.LlmRequest
import com.google.adk.kt.types.Content
import com.google.adk.kt.types.FinishReason
import com.google.adk.kt.types.FunctionCall
import com.google.adk.kt.types.FunctionResponse
import com.google.adk.kt.types.GenerateContentConfig
import com.google.adk.kt.types.Part
import com.google.adk.kt.types.Role
import com.google.adk.kt.types.Schema
import com.google.adk.kt.types.Tool
import com.google.adk.kt.types.Type
import com.google.adk.kt.types.FunctionDeclaration
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

class OllamaToolsTest : BehaviorSpec({
    Given("generation options") {
        When("derived from config") {
            Then("defaults are deterministic and config values are honored") {
                val defaults = generationOptions(GenerateContentConfig())
                defaults.getValue("num_ctx").jsonPrimitive.int shouldBe 32768
                defaults.getValue("temperature").jsonPrimitive.double shouldBe 0.0

                val honored = generationOptions(GenerateContentConfig(temperature = 0.5f, topP = 0.9f))
                honored.getValue("temperature").jsonPrimitive.double shouldBe 0.5
                (honored.getValue("top_p").jsonPrimitive.double in 0.89..0.91) shouldBe true
            }
        }
    }

    Given("the JSON-format switch") {
        When("inspecting the response MIME type") {
            Then("only an application/json type wants JSON") {
                wantsJSON(GenerateContentConfig()) shouldBe false
                wantsJSON(GenerateContentConfig(responseMimeType = "application/json")) shouldBe true
            }
        }
    }

    Given("a function declaration") {
        When("converting to Ollama tools") {
            Then("name, type, required and lowercased property types are projected") {
                val cfg =
                    GenerateContentConfig(
                        tools =
                            listOf(
                                Tool(
                                    functionDeclarations =
                                        listOf(
                                            FunctionDeclaration(
                                                name = "read_file",
                                                description = "read a repo file",
                                                parameters =
                                                    Schema(
                                                        type = Type.OBJECT,
                                                        properties = mapOf("path" to Schema(type = Type.STRING, description = "the path")),
                                                        required = listOf("path"),
                                                    ),
                                            ),
                                        ),
                                ),
                            ),
                    )
                val tools = toOllamaTools(cfg).shouldNotBeNull()
                tools.size shouldBe 1
                tools[0].jsonObject.getValue("type").jsonPrimitive.content shouldBe "function"
                val fn = tools[0].jsonObject.getValue("function").jsonObject
                fn.getValue("name").jsonPrimitive.content shouldBe "read_file"
                val params = fn.getValue("parameters").jsonObject
                params.getValue("type").jsonPrimitive.content shouldBe "object"
                params.getValue("required").jsonArray.size shouldBe 1
                val pathProp = params.getValue("properties").jsonObject.getValue("path").jsonObject
                pathProp.getValue("type").jsonArray[0].jsonPrimitive.content shouldBe "string"

                toOllamaTools(GenerateContentConfig()) shouldBe null
            }
        }
    }

    Given("a final response with a tool call") {
        When("building it") {
            Then("it is complete and carries text + function-call parts") {
                val tc = OllamaToolCall(OllamaToolCallFunction(name = "get_weather", arguments = buildJsonObject { put("city", "Paris") }))
                val resp = finalResponse("here you go", listOf(tc), "gemma4")
                resp.finishReason shouldBe FinishReason.STOP
                val parts = resp.content.shouldNotBeNull().parts
                parts.size shouldBe 2
                parts[0].text shouldBe "here you go"
                val fc = parts[1].functionCall.shouldNotBeNull()
                fc.name shouldBe "get_weather"
                fc.args["city"] shouldBe "Paris"

                finalResponse("", emptyList(), "gemma4").content.shouldNotBeNull().parts.size shouldBe 0
            }
        }
    }

    Given("a tool round-trip history") {
        When("flattening to Ollama messages") {
            Then("system, user, assistant tool-call and tool-result messages are produced") {
                val req =
                    LlmRequest(
                        config = GenerateContentConfig(systemInstruction = userText("sys")),
                        contents =
                            listOf(
                                Content(role = Role.USER, parts = listOf(Part(text = "read a.go"))),
                                Content(role = Role.MODEL, parts = listOf(Part(functionCall = FunctionCall(name = "read_file", args = mapOf("path" to "a.go"))))),
                                Content(role = Role.USER, parts = listOf(Part(functionResponse = FunctionResponse(name = "read_file", response = mapOf("content" to "package a"))))),
                            ),
                    )
                val msgs = toOllamaMessages(req)
                msgs.size shouldBe 4
                msgs[0].jsonObject.getValue("role").jsonPrimitive.content shouldBe "system"
                msgs[1].jsonObject.getValue("role").jsonPrimitive.content shouldBe "user"
                val asst = msgs[2].jsonObject
                asst.getValue("role").jsonPrimitive.content shouldBe "assistant"
                asst.getValue("tool_calls").jsonArray[0].jsonObject.getValue("function").jsonObject
                    .getValue("name").jsonPrimitive.content shouldBe "read_file"
                val toolMsg = msgs[3].jsonObject
                toolMsg.getValue("role").jsonPrimitive.content shouldBe "tool"
                toolMsg.getValue("tool_name").jsonPrimitive.content shouldBe "read_file"
                toolMsg.getValue("content").jsonPrimitive.content.contains("package a") shouldBe true
            }
        }
    }

    Given("a plain system/user/model history") {
        When("flattening to Ollama messages") {
            Then("each maps to the right role and content") {
                val req =
                    LlmRequest(
                        config = GenerateContentConfig(systemInstruction = userText("system rules")),
                        contents = listOf(userText("question"), assistantText("answer")),
                    )
                val msgs = toOllamaMessages(req)
                msgs.size shouldBe 3
                msgs.map { it.jsonObject.getValue("role").jsonPrimitive.content } shouldBe listOf("system", "user", "assistant")
                msgs.map { it.jsonObject.getValue("content").jsonPrimitive.content } shouldBe listOf("system rules", "question", "answer")
            }
        }
    }

    Given("an Ollama tool call") {
        When("converting to a genai function call") {
            Then("name and arguments are preserved") {
                val fc = toGenaiFunctionCall(OllamaToolCall(OllamaToolCallFunction(name = "f", arguments = buildJsonObject { put("x", "y") })))
                fc.name shouldBe "f"
                fc.args["x"] shouldBe "y"
            }
        }
    }

    Given("the jsonString helper") {
        When("serializing maps") {
            Then("null yields empty and a map yields compact JSON") {
                jsonString(null) shouldBe ""
                jsonString(mapOf("a" to "b")) shouldBe """{"a":"b"}"""
            }
        }
    }
})
